File: tree.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> visitedA(500009 , 0) ;
vector<int> visitedB(500009 , 0);
vector<int> adj[500009];
vector<int> s;
double ans = 0.0;

void initialize(){
    ans = 0.0;
    for(int i=0;i<500009;i++){
        visitedB[i]=0;
        visitedA[i]=0;
        adj[i].clear();
    }
}
void dfs(int node ,int a,int b){
    s.push_back(node);
    for(auto child: adj[node]){
        dfs(child , a,b);
    }
    visitedA[node]+=1;
    visitedB[node]+=1;
    s.pop_back();
    if(s.size() >=a)
        visitedA[s[s.size()-a]]+= visitedA[node];
    if(s.size() >=b)
        visitedB[s[s.size()-b]]+= visitedB[node];
    ans+=1.0 - (1.0 - ((double)visitedA[i])/n)*(1.0  - ((double)visitedB[i])/n);

}

int main() {
    int t;
    cin>>t;
    for(int z=1;z<=t;z++){
        initialize();
        int n,a,b;
        cin>>n>>a>>b;
        for(int i=0;i<n-1;i++){
        int c;  cin>>c;
        adj[c].push_back(i+2);
        }
        dfs(1,a,b);
        cout<<fixed;
        cout<<"Case #"<<z<<": "<<ans<<"\n";
    }
    return 0;
}

File: tree.kt
import java.util.*

fun main(args: Array<String>) {
    val t = readLine()!!.toInt()
    repeat(t) {
        val (n, a, b) = readLine()!!.split(' ').map { it.toInt() }
        val parents = readLine()!!.split(' ').map { it.toInt() }
        val graph = mutableMapOf<Int, MutableList<Int>>()
        for(i in 0 until n - 1){
            if (!graph.containsKey(parents[i])) graph[parents[i]] = mutableListOf()
            graph[parents[i]]!!.add(i + 2)
        }
        val tree = Tree(n, graph, a, b)
        println("Case #${it + 1}: ${tree.ans}")
    }
}

class Tree(val n : Int, val graph : MutableMap<Int, MutableList<Int>>, val a : Int, val b : Int){

    val visitedA = Array(n + 1) { 0 }
    val visitedB = Array(n + 1) { 0 }
    val path = Stack<Int>()
    var ans = 0.0

    init {
        dfs(1)
    }

    fun dfs(node : Int){
        path.push(node)
        if(graph.containsKey(node)) {
            graph[node]!!.forEach { dfs(it) }
        }
        visitedA[node]+= 1
        visitedB[node]+= 1
        path.pop()
        if(path.size >= a) visitedA[path[path.size - a]]+= visitedA[node]
        if(path.size >= b) visitedB[path[path.size - b]]+= visitedB[node]
        ans+= 1.0 - (1.0 - visitedA[node].toDouble()/n.toDouble())*(1.0 - visitedB[node].toDouble()/n.toDouble())
    }
}
